src/machine/machine-varlink.c

```diff
@@ -59,6 +59,12 @@ static int machine_leader(const char *name, sd_json_variant *variant, sd_json_di
         if (temp.pid == 1) /* refuse PID 1 */
                 return json_log(variant, flags, SYNTHETIC_ERRNO(EINVAL), "JSON field '%s' is not a valid leader PID.", strna(name));
 
+        if (!streq_ptr(name, "leaderPid") && pidref_is_set(leader)) {
+                if (!pidref_equal(leader, &temp))
+                        return json_log(variant, flags, SYNTHETIC_ERRNO(EINVAL), "JSON field '%s' conflicts with already dispatched leader PID.", strna(name));
+                return 0; /* honor PidRef obtained by leaderPid */
```

Comment:
I'd drop the conditionalization on `leaderPid`. if the two fields don't reference the same process it's a bug, no matter which one is listed first.